### PR TITLE
Support ipython directive

### PR DESCRIPTION
**Pull request by felix-hilden**
Related issue: #91

- [x] Tests written and passed
- [x] Documentation and changelog entry written, docs build passed
- [x] All `tox` checks passed

@OriolAbril could you review this, and particularly comment on your original regex. It had `\s*In\s*[...]\s`, so is there a situation where the notebook has multiple (or other) spaces there? I just used `^In [...] `.

@mgeier you've been involved with the other IPython stuff, so your review would be most helpful too!

### CODE REVIEW

**Comment by OriolAbril** [comment]
It is missing the ellipsis prefix (used for multiline inputs),  see for example https://github.com/arviz-devs/arviz/blob/4d63ca5d279374ee9b524aa9181a1fd5697a7423/arviz/stats/stats.py#L143.

As for the spaces, I haven't tried but assumed an extra space between the In and the number would work too

File: tests/extension/ref/ref_ipython.txt [hunk 1]
```diff
@@ -5,10 +5,14 @@ test_project.bar
...
-   %cd subdir
-   import test_project
-   test_project.bar()
+   In [1]: %cd subdir
+   In [2]: import test_project
+   In [3]: b = test_project.bar()
+   In [4]: 2 + 2
+   Out[4]: 4
+   In [5]: class A:
+      ...:     pass
```
Comment (felix-hilden) [inline tests/extension/ref/ref_ipython.txt:16]: @OriolAbril I've tested the ellipsis here. The parsing works such that the sanitation is done by the `TransformerManager`. We just decide to erase everything after `Out` prefixes!

**Comment by mgeier** [comment]
This works very well!

... except for a few corner cases ...

I've tried it with this file: https://github.com/ipython/ipython/blob/master/docs/source/sphinxext.rst

There are several errors due to (intentional) syntax errors in the examples.

I think it would be better to use `logger.warning()` instead of `logger.error()` for syntax errors, as this would be less disruptive for users.

And there is an unhandled `IndentationError` when using this example:

```rst
.. ipython::
   :verbatim:

   In [15]: !wc *
       2    12    77 README.txt
      40    97   884 buttons.py
      26    90   712 check_buttons.py
      19    52   416 cursor.py
     180   404  4882 menu.py
      16    45   337 multicursor.py
      36   106   916 radio_buttons.py
      48   226  2082 rectangle_selector.py
      43   118  1063 slider_demo.py
      40   124  1088 span_selector.py
     450  1274 12457 total
```

I think this should also be turned into a Sphinx warning.

Maybe the `@print_exceptions` decorator could be turned into `@turn_exceptions_into_sphinx_warnings` or something like this?

**Comment by felix-hilden** [comment]
Good points! I think we'll add `IndentationError` to the handler with `SyntaxError` because it is similar in intent. And I agree that `logger.warning` could be better there. Just to be sure: you agree that the intentional syntax errors should produce warnings, right? These are more general than this PR though, so I'll make another issue.

I don't want to change the decorator though, because it now only catches unexpected errors to print their stack traces regardless of Sphinx, so that debugging is easier. Any errors should definitely disrupt the build.

**Comment by felix-hilden** [comment]
Oh, the error happens in *transforming* the block for parsing! So we'll have to add an error handler there.

**Comment by mgeier** [comment]
> Just to be sure: you agree that the intentional syntax errors should produce warnings, right?

Yes.

The problem now is that the `location=node` thing doesn't seem to work with the new warnings.
We always get line number 1 instead of the correct line number.

**Comment by felix-hilden** [comment]
Good catch! It seems that the location works fine overall (e.g. `.. code:: ipython` works), but `node.line` for whatever the IPython directive inserts is 1. This only happens if `:verbatim:` is used though, because otherwise IPython itself catches the syntax error. I'm inclined to chuck this onto IPython, since the directive is still in beta!

**Pull request merged**